File: python/src/vigilai/monitoring_agent.py
"""Monitoring agent for collecting metrics and errors"""

import os
import time
import psutil
import threading
import traceback
import random
from typing import Optional, Dict, Any, List
from .types import Metric, ErrorEvent, MonitoringData, VigilAIConfig
from .api_client import APIClient
from .redactor import Redactor


class CircularBuffer:
    """Circular buffer for storing data with fixed capacity"""
    
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.buffer: List[Any] = []
        self.head = 0
        self.size = 0
    
    def push(self, item: Any) -> None:
        """Add item to buffer, overwriting oldest if full"""
        if self.size < self.capacity:
            self.buffer.append(item)
            self.size += 1
        else:
            self.buffer[self.head] = item
        self.head = (self.head + 1) % self.capacity
    
    def get_all(self) -> List[Any]:
        """Get all items in chronological order"""
        if self.size < self.capacity:
            return self.buffer[:self.size]
        # Return items in order: from head to end, then from start to head
        return self.buffer[self.head:] + self.buffer[:self.head]
    
    def clear(self) -> None:
        """Clear all items from buffer"""
        self.buffer = []
        self.head = 0
        self.size = 0
    
    def get_size(self) -> int:
        """Get current number of items in buffer"""
        return self.size
    
    def is_full(self) -> bool:
        """Check if buffer is at capacity"""
        return self.size >= self.capacity


class MonitoringAgent:
    """Monitoring agent responsible for collecting metrics and errors"""
    
    def __init__(self, config: VigilAIConfig, api_client: APIClient):
        self.config = config
        self.api_client = api_client
        self.redactor = Redactor(
            config.security.redaction_rules,
            config.security.enable_pii_redaction
        )
        self.metrics_buffer = CircularBuffer(config.monitoring.buffer_size)
        self.errors_buffer = CircularBuffer(config.monitoring.buffer_size)
        self.monitoring_thread: Optional[threading.Thread] = None
        self.flush_thread: Optional[threading.Thread] = None
        self.monitoring_active = False
        self.last_flush_time = time.time()
        self.request_count = 0
        self.last_request_count = 0
        self.last_throughput_check = time.time()
        self.process = psutil.Process(os.getpid())
        self.transmission_attempts = 0
        self.transmission_successes = 0
        self.transmission_failures = 0
        self.last_transmission_error: Optional[str] = None
        self.last_transmission_success: Optional[int] = None
    
    def start_monitoring(self) -> None:
        """Start monitoring system metrics"""
        if self.monitoring_active:
            return  # Already monitoring
        
        self.monitoring_active = True
        
        # Collect metrics immediately on start
        self.collect_system_metrics()
        
        self.monitoring_thread = threading.Thread(target=self._monitoring_loop, daemon=True)
        self.monitoring_thread.start()
        
        # Start flush thread to check for time-based flushes
        self.flush_thread = threading.Thread(target=self._flush_loop, daemon=True)
        self.flush_thread.start()
    
    def stop_monitoring(self) -> None:
        """Stop monitoring"""
        self.monitoring_active = False
        if self.monitoring_thread:
            self.monitoring_thread.join(timeout=2.0)
            self.monitoring_thread = None
        if self.flush_thread:
            self.flush_thread.join(timeout=2.0)
            self.flush_thread = None
    
    def _monitoring_loop(self) -> None:
        """Background loop for collecting system metrics"""
        interval_seconds = self.config.monitoring.interval / 1000.0
        
        while self.monitoring_active:
            self.collect_system_metrics()
            time.sleep(interval_seconds)
    
    def _flush_loop(self) -> None:
        """Background loop for checking flush conditions"""
        interval_seconds = self.config.monitoring.interval / 1000.0
        
        while self.monitoring_active:
            if self.should_flush():
                try:
                    self.transmit_data()
                except Exception as e:
                    # Log error but don't throw - monitoring should continue
                    print(f"Failed to transmit data: {str(e)}")
            time.sleep(interval_seconds)
    
    def capture_request(
        self,
        endpoint: str,
        method: str,
        status_code: int,
        duration: float
    ) -> None:
        """Capture HTTP request metrics"""
        self.request_count += 1
        
        # Apply sampling rate
        if random.random() > self.config.monitoring.sampling_rate:
            return
        
        timestamp = int(time.time() * 1000)
        
        # Capture response time metric
        self.metrics_buffer.push(Metric(
            name='http.response_time',
            value=duration,
            timestamp=timestamp,
            tags={
                'endpoint': endpoint,
                'method': method,
                'status': str(status_code),
            }
        ))
        
        # Capture status code metric
        self.metrics_buffer.push(Metric(
            name='http.status',
            value=float(status_code),
            timestamp=timestamp,
            tags={
                'endpoint': endpoint,
                'method': method,
            }
        ))
        
        # Check if buffer should be flushed due to capacity
        if self.should_flush():
            try:
                self.transmit_data()
            except Exception as e:
                print(f"Failed to transmit data: {str(e)}")
    
    def capture_error(self, error: Exception, context: Optional[Dict[str, Any]] = None) -> None:
        """Capture error with stack trace"""
        error_event = ErrorEvent(
            message=str(error),
            stack=traceback.format_exc(),
            timestamp=int(time.time() * 1000),
            context=context
        )
        
        self.errors_buffer.push(error_event)
        
        # Check if buffer should be flushed due to capacity
        if self.should_flush():
            try:
                self.transmit_data()
            except Exception as e:
                print(f"Failed to transmit data: {str(e)}")
    
    def collect_system_metrics(self) -> None:
        """Collect system metrics (CPU, memory, throughput)"""
        timestamp = int(time.time() * 1000)
        
        # Memory usage
        memory_info = self.process.memory_info()
        self.metrics_buffer.push(Metric(
            name='system.memory.rss',
            value=round(memory_info.rss / 1024 / 1024, 2),  # Convert to MB
            timestamp=timestamp
        ))
        
        # CPU usage (percentage)
        try:
            cpu_percent = self.process.cpu_percent(interval=0.1)
            self.metrics_buffer.push(Metric(
                name='system.cpu.usage',
                value=round(cpu_percent, 2),
                timestamp=timestamp
            ))
        except Exception:
            # cpu_percent can fail on first call, ignore
            pass
        
        # Request throughput (requests per second)
        now = time.time()
        time_diff = now - self.last_throughput_check
        if time_diff > 0:
            throughput = (self.request_count - self.last_request_count) / time_diff
            self.metrics_buffer.push(Metric(
                name='http.throughput',
                value=round(throughput, 2),
                timestamp=timestamp
            ))
            
            self.last_request_count = self.request_count
            self.last_throughput_check = now
    
    def get_buffered_data(self) -> MonitoringData:
        """Get all buffered data, filtering out data older than retention period"""
        now = int(time.time() * 1000)
        retention_cutoff = now - self.config.security.data_retention_period
        
        all_metrics = self.metrics_buffer.get_all()
        all_errors = self.errors_buffer.get_all()
        
        return MonitoringData(
            metrics=[m for m in all_metrics if m.timestamp >= retention_cutoff],
            errors=[e for e in all_errors if e.timestamp >= retention_cutoff]
        )
    
    def clear_buffer(self) -> None:
        """Clear all buffered data"""
        self.metrics_buffer.clear()
        self.errors_buffer.clear()
        self.last_flush_time = time.time()
    
    def should_flush(self) -> bool:
        """Check if buffer should be flushed"""
        time_since_flush = (time.time() - self.last_flush_time) * 1000  # Convert to ms
        time_threshold = self.config.monitoring.interval
        
        return (
            self.metrics_buffer.is_full() or
            self.errors_buffer.is_full() or
            time_since_flush >= time_threshold
        )
    
    def transmit_data(self) -> None:
        """Transmit buffered data to backend in batches"""
        data = self.get_buffered_data()
        
        # Only transmit if there's data
        if len(data.metrics) == 0 and len(data.errors) == 0:
            return
        
        # Batch size limit (up to 100 events per transmission)
        BATCH_SIZE = 100
        total_events = len(data.metrics) + len(data.errors)
        
        if total_events <= BATCH_SIZE:
            # Single batch transmission
            self._transmit_batch(data)
        else:
            # Multiple batch transmissions
            metrics_offset = 0
            errors_offset = 0
            
            while metrics_offset < len(data.metrics) or errors_offset < len(data.errors):
                metrics_in_batch = min(BATCH_SIZE, len(data.metrics) - metrics_offset)
                remaining_space = BATCH_SIZE - metrics_in_batch
                errors_in_batch = min(remaining_space, len(data.errors) - errors_offset)
                
                batch = MonitoringData(
                    metrics=data.metrics[metrics_offset:metrics_offset + metrics_in_batch],
                    errors=data.errors[errors_offset:errors_offset + errors_in_batch]
                )
                
                self._transmit_batch(batch)
                
                metrics_offset += metrics_in_batch
                errors_offset += errors_in_batch
        
        # Clear buffer after all batches transmitted successfully
        self.clear_buffer()
    
    def _transmit_batch(self, data: MonitoringData) -> None:
        """Transmit a single batch of data"""
        self.transmission_attempts += 1
        
        try:
            # Apply redaction before transmission
            redacted_data = MonitoringData(
                metrics=self.redactor.redact_metrics([
                    {
                        'name': m.name,
                        'value': m.value,
                        'timestamp': m.timestamp,
                        'tags': m.tags
                    } for m in data.metrics
                ]),
                errors=self.redactor.redact_errors([
                    {
                        'message': e.message,
                        'stack': e.stack,
                        'timestamp': e.timestamp,
                        'context': e.context
                    } for e in data.errors
                ])
            )
            
            self.api_client.transmit_monitoring_data(redacted_data)
            self.transmission_successes += 1
            self.last_transmission_success = int(time.time() * 1000)
        except Exception as e:
            self.transmission_failures += 1
            self.last_transmission_error = str(e)
            raise
    
    def get_buffer_stats(self) -> Dict[str, int]:
        """Get buffer statistics"""
        return {
            'metrics_count': self.metrics_buffer.get_size(),
            'errors_count': self.errors_buffer.get_size(),
        }
    
    def is_monitoring(self) -> bool:
        """Check if monitoring is active"""
        return self.monitoring_active
    
    def get_transmission_stats(self) -> Dict[str, Any]:
        """Get transmission statistics"""
        success_rate = (
            self.transmission_successes / self.transmission_attempts
            if self.transmission_attempts > 0
            else 1.0
        )
        
        stats = {
            'attempts': self.transmission_attempts,
            'successes': self.transmission_successes,
            'failures': self.transmission_failures,
            'success_rate': success_rate,
        }
        
        if self.last_transmission_error:
            stats['last_error'] = self.last_transmission_error
        if self.last_transmission_success:
            stats['last_success'] = self.last_transmission_success
        
        return stats

File: typescript/src/monitoring-agent.ts
import * as os from 'os';
import { Metric, ErrorEvent, MonitoringData, ResolvedConfig } from './types';
import { APIClient } from './api-client';
import { Redactor } from './redactor';

/**
 * Circular buffer for storing monitoring data
 */
class CircularBuffer<T> {
  private buffer: T[];
  private head: number = 0;
  private size: number = 0;
  private readonly capacity: number;

  constructor(capacity: number) {
    this.capacity = capacity;
    this.buffer = new Array(capacity);
  }

  push(item: T): void {
    this.buffer[this.head] = item;
    this.head = (this.head + 1) % this.capacity;
    if (this.size < this.capacity) {
      this.size++;
    }
  }

  getAll(): T[] {
    if (this.size < this.capacity) {
      return this.buffer.slice(0, this.size);
    }
    // Return items in order: from head to end, then from start to head
    return [...this.buffer.slice(this.head), ...this.buffer.slice(0, this.head)];
  }

  clear(): void {
    this.head = 0;
    this.size = 0;
  }

  getSize(): number {
    return this.size;
  }

  isFull(): boolean {
    return this.size >= this.capacity;
  }
}

/**
 * Monitoring agent responsible for collecting metrics and errors
 */
export class MonitoringAgent {
  private metricsBuffer: CircularBuffer<Metric>;
  private errorsBuffer: CircularBuffer<ErrorEvent>;
  private config: ResolvedConfig;
  private apiClient: APIClient;
  private redactor: Redactor;
  private monitoringInterval?: NodeJS.Timeout;
  private flushInterval?: NodeJS.Timeout;
  private lastFlushTime: number = Date.now();
  private requestCount: number = 0;
  private lastRequestCount: number = 0;
  private lastThroughputCheck: number = Date.now();
  private transmissionAttempts: number = 0;
  private transmissionSuccesses: number = 0;
  private transmissionFailures: number = 0;
  private lastTransmissionError?: string;
  private lastTransmissionSuccess?: number;

  constructor(config: ResolvedConfig, apiClient: APIClient) {
    this.config = config;
    this.apiClient = apiClient;
    this.redactor = new Redactor(
      config.security.redactionRules,
      config.security.enablePIIRedaction
    );
    this.metricsBuffer = new CircularBuffer<Metric>(config.monitoring.bufferSize);
    this.errorsBuffer = new CircularBuffer<ErrorEvent>(config.monitoring.bufferSize);
  }

  /**
   * Start monitoring system metrics
   */
  startMonitoring(): void {
    if (this.monitoringInterval) {
      return; // Already monitoring
    }

    // Collect metrics immediately on start
    this.collectSystemMetrics();

    this.monitoringInterval = setInterval(() => {
      this.collectSystemMetrics();
    }, this.config.monitoring.interval);

    // Start flush interval to check for time-based flushes
    this.flushInterval = setInterval(() => {
      if (this.shouldFlush()) {
        this.transmitData().catch((error) => {
          // Log error but don't throw - monitoring should continue
          console.error('Failed to transmit data:', error instanceof Error ? error.message : 'Unknown error');
        });
      }
    }, this.config.monitoring.interval);
  }

  /**
   * Stop monitoring
   */
  stopMonitoring(): void {
    if (this.monitoringInterval) {
      clearInterval(this.monitoringInterval);
      this.monitoringInterval = undefined;
    }
    if (this.flushInterval) {
      clearInterval(this.flushInterval);
      this.flushInterval = undefined;
    }
  }

  /**
   * Capture HTTP request metrics
   */
  captureRequest(
    endpoint: string,
    method: string,
    statusCode: number,
    duration: number
  ): void {
    this.requestCount++;

    // Apply sampling rate
    if (Math.random() > this.config.monitoring.samplingRate) {
      return;
    }

    const timestamp = Date.now();

    // Capture response time metric
    this.metricsBuffer.push({
      name: 'http.response_time',
      value: duration,
      timestamp,
      tags: {
        endpoint,
        method,
        status: statusCode.toString(),
      },
    });

    // Capture status code metric
    this.metricsBuffer.push({
      name: 'http.status',
      value: statusCode,
      timestamp,
      tags: {
        endpoint,
        method,
      },
    });

    // Check if buffer should be flushed due to capacity
    if (this.shouldFlush()) {
      this.transmitData().catch((error) => {
        console.error('Failed to transmit data:', error instanceof Error ? error.message : 'Unknown error');
      });
    }
  }

  /**
   * Capture error with stack trace
   */
  captureError(error: Error, context?: Record<string, any>): void {
    const errorEvent: ErrorEvent = {
      message: error.message,
      stack: error.stack || '',
      timestamp: Date.now(),
      context,
    };

    this.errorsBuffer.push(errorEvent);

    // Check if buffer should be flushed due to capacity
    if (this.shouldFlush()) {
      this.transmitData().catch((error) => {
        console.error('Failed to transmit data:', error instanceof Error ? error.message : 'Unknown error');
      });
    }
  }

  /**
   * Collect system metrics (CPU, memory, throughput)
   */
  private collectSystemMetrics(): void {
    const timestamp = Date.now();

    // Memory usage
    const memoryUsage = process.memoryUsage();
    this.metricsBuffer.push({
      name: 'system.memory.heap_used',
      value: Math.round(memoryUsage.heapUsed / 1024 / 1024), // Convert to MB
      timestamp,
    });

    this.metricsBuffer.push({
      name: 'system.memory.heap_total',
      value: Math.round(memoryUsage.heapTotal / 1024 / 1024), // Convert to MB
      timestamp,
    });

    // CPU usage (approximation using os.loadavg)
    const loadAvg = os.loadavg();
    const cpuCount = os.cpus().length;
    const cpuUsagePercent = (loadAvg[0] / cpuCount) * 100;

    this.metricsBuffer.push({
      name: 'system.cpu.usage',
      value: Math.round(cpuUsagePercent * 100) / 100, // Round to 2 decimal places
      timestamp,
    });

    // Request throughput (requests per second)
    const now = Date.now();
    const timeDiff = (now - this.lastThroughputCheck) / 1000; // Convert to seconds
    if (timeDiff > 0) {
      const throughput = (this.requestCount - this.lastRequestCount) / timeDiff;
      this.metricsBuffer.push({
        name: 'http.throughput',
        value: Math.round(throughput * 100) / 100, // Round to 2 decimal places
        timestamp,
      });

      this.lastRequestCount = this.requestCount;
      this.lastThroughputCheck = now;
    }
  }

  /**
   * Get all buffered data, filtering out data older than retention period
   */
  getBufferedData(): MonitoringData {
    const now = Date.now();
    const retentionCutoff = now - this.config.security.dataRetentionPeriod;
    
    return {
      metrics: this.metricsBuffer.getAll().filter(m => m.timestamp >= retentionCutoff),
      errors: this.errorsBuffer.getAll().filter(e => e.timestamp >= retentionCutoff),
    };
  }

  /**
   * Clear all buffered data
   */
  clearBuffer(): void {
    this.metricsBuffer.clear();
    this.errorsBuffer.clear();
    this.lastFlushTime = Date.now();
  }

  /**
   * Check if buffer should be flushed
   */
  shouldFlush(): boolean {
    const timeSinceFlush = Date.now() - this.lastFlushTime;
    const timeThreshold = this.config.monitoring.interval;

    return (
      this.metricsBuffer.isFull() ||
      this.errorsBuffer.isFull() ||
      timeSinceFlush >= timeThreshold
    );
  }

  /**
   * Transmit buffered data to backend in batches
   */
  async transmitData(): Promise<void> {
    const data = this.getBufferedData();
    
    // Only transmit if there's data
    if (data.metrics.length === 0 && data.errors.length === 0) {
      return;
    }

    // Batch size limit (up to 100 events per transmission)
    const BATCH_SIZE = 100;
    const totalEvents = data.metrics.length + data.errors.length;
    
    if (totalEvents <= BATCH_SIZE) {
      // Single batch transmission
      await this.transmitBatch(data);
    } else {
      // Multiple batch transmissions
      let metricsOffset = 0;
      let errorsOffset = 0;
      
      while (metricsOffset < data.metrics.length || errorsOffset < data.errors.length) {
        const metricsInBatch = Math.min(BATCH_SIZE, data.metrics.length - metricsOffset);
        const remainingSpace = BATCH_SIZE - metricsInBatch;
        const errorsInBatch = Math.min(remainingSpace, data.errors.length - errorsOffset);
        
        const batch: MonitoringData = {
          metrics: data.metrics.slice(metricsOffset, metricsOffset + metricsInBatch),
          errors: data.errors.slice(errorsOffset, errorsOffset + errorsInBatch),
        };
        
        await this.transmitBatch(batch);
        
        metricsOffset += metricsInBatch;
        errorsOffset += errorsInBatch;
      }
    }
    
    // Clear buffer after all batches transmitted successfully
    this.clearBuffer();
  }

  /**
   * Transmit a single batch of data
   */
  private async transmitBatch(data: MonitoringData): Promise<void> {
    this.transmissionAttempts++;
    
    try {
      // Apply redaction before transmission
      const redactedData: MonitoringData = {
        metrics: this.redactor.redactMetrics(data.metrics),
        errors: this.redactor.redactErrors(data.errors),
      };
      
      await this.apiClient.transmitMonitoringData(redactedData);
      this.transmissionSuccesses++;
      this.lastTransmissionSuccess = Date.now();
    } catch (error) {
      this.transmissionFailures++;
      this.lastTransmissionError = error instanceof Error ? error.message : 'Unknown error';
      throw error;
    }
  }

  /**
   * Get buffer statistics
   */
  getBufferStats(): { metricsCount: number; errorsCount: number } {
    return {
      metricsCount: this.metricsBuffer.getSize(),
      errorsCount: this.errorsBuffer.getSize(),
    };
  }

  /**
   * Check if monitoring is active
   */
  isMonitoring(): boolean {
    return this.monitoringInterval !== undefined;
  }

  /**
   * Get transmission statistics
   */
  getTransmissionStats(): {
    attempts: number;
    successes: number;
    failures: number;
    successRate: number;
    lastError?: string;
    lastSuccess?: number;
  } {
    const successRate = this.transmissionAttempts > 0
      ? this.transmissionSuccesses / this.transmissionAttempts
      : 1.0;

    return {
      attempts: this.transmissionAttempts,
      successes: this.transmissionSuccesses,
      failures: this.transmissionFailures,
      successRate,
      lastError: this.lastTransmissionError,
      lastSuccess: this.lastTransmissionSuccess,
    };
  }
}
